File: internal/indexer/testdata/rust/services/core_model_service.rs
use crate::models::core_model::CoreModel;

/// CoreModelRepository stores and retrieves CoreModel instances.
/// Demonstrates: use-import (0.40), struct-field composition (0.25),
/// CoreModel::new instantiation (0.95 via lexical), method calls (0.65).
pub struct CoreModelRepository {
    store: Vec<CoreModel>,
}

impl CoreModelRepository {
    /// Creates a new repository seeded with an initial CoreModel.
    /// Coupling: CoreModel::new → instantiation (lexical escalation → 0.95).
    pub fn new() -> CoreModelRepository {
        let initial = CoreModel::new("initial", 0.5);
        CoreModelRepository { store: vec![initial] }
    }

    /// add stores a CoreModel in the repository.
    pub fn add(&mut self, model: CoreModel) {
        if model.is_valid() {
            self.store.push(model);
        }
    }

    /// find_by_title returns the first CoreModel matching the title.
    pub fn find_by_title(&self, title: &str) -> Option<&CoreModel> {
        self.store.iter().find(|m| m.title == title)
    }

    /// describe calls describe() on each CoreModel.
    pub fn describe(&self) -> Vec<&str> {
        self.store.iter().map(|m| m.describe()).collect()
    }
}

/// store_core_model persists a CoreModel using a standalone function.
pub fn store_core_model(model: CoreModel) -> bool {
    model.is_valid()
}

/// build_core_model uses the CoreModel constructor — another instantiation pattern.
pub fn build_core_model(title: &str, score: f64) -> CoreModel {
    CoreModel::new(title, score)
}

File: internal/indexer/testdata/go/services/core_model_service.go
package services

import "models"

// CoreModelRepository stores and retrieves CoreModel instances.
// Demonstrates instantiation (high coupling) and method call (medium coupling).
type CoreModelRepository struct {
	store []models.CoreModel
}

// NewCoreModelRepository creates a populated repository with two CoreModel entries.
// Coupling: struct literal instantiation → high-coupling usage of CoreModel.
func NewCoreModelRepository() *CoreModelRepository {
	first := &models.CoreModel{Title: "alpha", Score: 1.0}
	second := models.CoreModel{Title: "beta", Score: 0.5}
	return &CoreModelRepository{
		store: []models.CoreModel{*first, second},
	}
}

// Add stores a CoreModel by instantiating from raw fields.
// Coupling: struct literal + method call → high coupling.
func (r *CoreModelRepository) Add(title string, score float64) {
	m := &models.CoreModel{Title: title, Score: score}
	if m.IsValid() {
		r.store = append(r.store, *m)
	}
}

// Find returns a CoreModel matching the title.
func (r *CoreModelRepository) Find(title string) *models.CoreModel {
	for i := range r.store {
		if r.store[i].Title == title {
			return &r.store[i]
		}
	}
	return nil
}

// Describe calls Describe on each CoreModel and returns the first non-empty result.
// Coupling: method call → medium coupling.
func (r *CoreModelRepository) Describe() string {
	for i := range r.store {
		if d := r.store[i].Describe(); d != "" {
			return d
		}
	}
	return ""
}

// StoreCoreModel persists a CoreModel using a standalone function.
func StoreCoreModel(m models.CoreModel) bool {
	return m.IsValid()
}

// LoadCoreModel fabricates a CoreModel by value — another instantiation.
func LoadCoreModel(title string) models.CoreModel {
	return models.CoreModel{Title: title, Score: 0.0}
}
